### PR TITLE
[RFC 140 2a] `pkgs/by-name`: Enforce for new packages

### PR DESCRIPTION
**Pull request by infinisil**
> [!note]
> An open call for the final review and merge was held on 2024-01-15, see https://github.com/NixOS/nixpkgs/pull/275539#issuecomment-1892513546

## Context

This is part of the effort to implement [RFC 140](https://github.com/NixOS/rfcs/pull/140), aka `pkgs/by-name`. While [previous PRs](https://github.com/NixOS/nixpkgs/pull/237439) already enabled users to define packages using `pkgs/by-name`, there's still thousands of packages still using `all-packages.nix`, and new ones added [every day](https://github.com/NixOS/nixpkgs/commits/master/pkgs/top-level/all-packages.nix).

This PR is part of the [RFC 140 migration plan](https://github.com/NixOS/nixpkgs/issues/258650), building on top of https://github.com/NixOS/nixpkgs/pull/272395 and https://github.com/NixOS/nixpkgs/pull/274591.

This work is sponsored by [Antithesis](https://antithesis.com/) and [Tweag](https://www.tweag.io/) :sparkles:

## Changes

This PR prevents _new_ packages using `pkgs.callPackage` from being defined not in `pkgs/by-name`. Instead the `pkgs/by-name` directory has to be used instead. This the A-side of [Part 2 of the RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-2-automated-migration).

This doesn't affect packages defined using alternate `callPackage`'s like `libsForQt5.callPackage` or `callPackages`, which `pkgs/by-name` cannot handle (yet?).

> [!Note]
> The effects of this PR will only be seen once the nixos-unstable channel updates, since its pre-built version of the tooling is used for CI.

## Things done

- [x] Tests updated
  - [x] Test to ensure that new packages using `callPackage` but not using `pkgs/by-name` throw an error
  - [x] Test to ensure that a package moved out of `pkgs/by-name` throws an error
  - [x] Test to ensure that new packages using an alternate `callPackage` doesn't give an error.
- [x] Tested on the entirety of Nixpkgs
  - [x] Ensured okay performance (about 25 single-core seconds on my machine)
- [x] Tidy code
- [x] Clean history

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by infinisil** [comment]
This PR is now theoretically ready, but since the changes got a bit big, I split off everything not directly related to this PR into https://github.com/NixOS/nixpkgs/pull/278805, which should be merged first.

File: pkgs/by-name/README.md
Comment (Ericson2314) [inline pkgs/by-name/README.md:8]: I think it is still better to lead with the positive part. Something like

> Packages found in the name-based structure are automatically included, without needing to be added in a file like `all-packages.nix`. They must only occur in file if they require overriding the default value of an implicit attribute (see below).

File: pkgs/by-name/README.md
Comment (philiptaron) [inline pkgs/by-name/README.md:4]: Add new top-level packages to Nixpkgs using this mechanism [whenever possible](#limitations).

File: pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:245]: Comma splice in the last sentence. Replace comma with semicolon.

> This should not be necessary; please move the package back and remove the manual `callPackage`.

File: pkgs/test/nixpkgs-check-by-name/src/ratchet.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/ratchet.rs:37]: This name is super general; I mildly prefer "Package" and I'm interested in what motivated the switch.

File: pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:263]: For my notes: [pkgs/test/nixpkgs-check-by-name/tests/new-package-non-by-name/expected](https://github.com/NixOS/nixpkgs/pull/275539/files#diff-1c1e6bc11bf7b0b34e539456211d57c10e9418043fda5d9c3e512a5ee96c4a51) is the test of this.

File: pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs:273]: Comma splice, add period to avoid.

This is a new top-level package of the form `callPackage {call_package_arg} {{ }}`. Define it in {}. See `pkgs/by-name/README.md` for more details.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:279]: Docs claim it's sorted; now it's in the order that it was decoded.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:204]: Future ratchets might try to drive this number down, I might imagine.

File: pkgs/test/nixpkgs-check-by-name/src/eval.nix
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.nix:112]: I don't think this is sorted anymore. Might not matter.

**Review by philiptaron — APPROVED** [review]
I took a thorough look. The tests are quite clear. Nice work. I don't have any code related comments, just copy-editing. Might also run a RFC101 formatter over these Nix files just for giggles.

File: pkgs/test/nixpkgs-check-by-name/src/ratchet.rs
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/ratchet.rs:37]: Oh yeah it's actually still just packages here. I switched the `eval.rs` code from packages -> attributes because it now handles non-package attributes, but that doesn't apply here, will changed back!

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:204]: Oh yeah, I thought about implementing an eval-successful ratchet, but I think this should be a future PR. Then also it's also not `pkgs/by-name`-specific anymore and the tool could use a rename.

File: pkgs/test/nixpkgs-check-by-name/src/eval.nix
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.nix]: Doesn't matter much, but it makes the resulting error ordering a bit nicer, I adjusted it now to always sort it in https://github.com/NixOS/nixpkgs/pull/275539/commits/cbff02834f3c0a1daa3502566fcf19433f9ca214 :)

File: pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs]: Semicolon feels a bit weird, I just removed the "This should not be necessary" part. That part of the sentence itself shouldn't be necessary. If we discover cases where it's still necessary we can reconsider

File: pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/nixpkgs_problem.rs]: Done now :)

File: pkgs/by-name/README.md
Comment (infinisil) [reply pkgs/by-name/README.md]: I changed the wording a little bit, but this is essentially applied now :)

**Comment by infinisil** [comment]
This PR is now scheduled for a final review and ideal merge on Monday, 2024-01-15 17:00 CET (date -d@1705334400). Feel free to join! https://meet.jit.si/nixpkgs-architecture

Of course if any blocking feedback comes up until then I won't merge it until addressed.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs
Comment (philiptaron) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:204]: One only needs to look at the list of things detected by NixpkgsProblems to realize we've backed ourselves into a damn fine nixpkgs linter framework. 🚀

**Review by philiptaron — APPROVED** [review]
Ship it.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs
Comment (Ericson2314) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:279]: This is because we have the non-by-name names now too? Can we get `package_names` out of scope or something by this point?

Comment (Ericson2314) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:279]: I just mean the fact that `package_names` the parameter is in scope but we do *not* want to use it to define `package_names` the field is a bit intuitive. If `package_names` the parameter had gone out of scope so we had to use a more complicated expression (or one of the two identifiers would be renamed) it would be a bit more intuitive.

Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:279]: Ah I see. I don't think it's a big problem really, it's really just potentially one extra line of change in the future in case it would be renamed. And it wouldn't cause any code breakage.

Comment (Ericson2314) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:279]: Yeah it's just a small nit. No a merge blocker.

**Comment by infinisil** [comment]
We looked at this together today in a call with @tomberek @Ericson2314 @philiptaron @roberth and Dominic. We went through the context, the code and the review comments and couldn't find anything blocking. We agreed to proceed to the merge! :rocket:

**Review by roberth — APPROVED** [review]

**Review by Ericson2314 — APPROVED** [review]

**Comment by infinisil** [comment]
Noticed CI failing in the last second, now fixed, see commit message. I'll merge once CI passes!

**Review by philiptaron — APPROVED** [review]

**Comment by infinisil** [comment]
Now that nixos-unstable contains the latest fixes, here's the final follow-up PR: https://github.com/NixOS/nixpkgs/pull/281835 which starts enabling the enforcement!

**Pull request merged**
